packages/grid/src/vaadin-grid-keyboard-navigation-mixin.js

```diff
@@ -729,23 +729,14 @@ export const KeyboardNavigationMixin = (superClass) =>
         // The focus is about to exit the grid to the bottom.
         this.$.focusexit.focus();
       } else if (focusTarget === this._itemsFocusable) {
-        let itemsFocusTarget = focusTarget;
-        const targetRow = isRow(focusTarget) ? focusTarget : focusTarget.parentNode;
-        this._ensureScrolledToIndex(this._focusedItemIndex);
-        if (targetRow.index !== this._focusedItemIndex && isCell(focusTarget)) {
-          // The target row, which is about to be focused next, has been
-          // assigned with a new index since last focus, probably because of
-          // scrolling. Focus the row for the stored focused item index instead.
-          const columnIndex = Array.from(targetRow.children).indexOf(this._itemsFocusable);
-          const focusedItemRow = Array.from(this.$.items.children).find(
-            (row) => !row.hidden && row.index === this._focusedItemIndex,
-          );
-          if (focusedItemRow) {
-            itemsFocusTarget = focusedItemRow.children[columnIndex];
-          }
-        }
+        this.__ensureFlatIndexInViewport(this._focusedItemIndex);
+
+        // Ensure the correct element is set as focusable after scrolling.
+        // The virtualizer may use a different element to render the item.
+        this.__updateItemsFocusable();
```

Comment:
**note:** The previous version didn't update the focusable element if it was a row, which sometimes resulted in setting the focus outline on an incorrect row.